File: backend/routes/admin/testimonials.ts
import { Router, Request, Response } from 'express';
import db from '../../config/database';
import { deleteFromS3 } from '../../config/s3';

const router = Router();

const ALLOWED_PLATFORMS = new Set([
  'tiktok', 'zalo', 'instagram', 'other',
]);

function normalisePlatform(p: unknown): string {
  if (typeof p === 'string' && ALLOWED_PLATFORMS.has(p)) return p;
  return 'other';
}

// GET /api/admin/testimonials
router.get('/', async (_req: Request, res: Response) => {
  try {
    const result = await db.query(
      `SELECT * FROM testimonials
       ORDER BY is_featured DESC, sort_order ASC, created_at DESC`,
    );
    return res.json({ success: true, testimonials: result.rows });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// POST /api/admin/testimonials
// Body: { image_url, platform?, reviewer_name?, caption?, is_featured?, is_featured_on_home? }
router.post('/', async (req: Request, res: Response) => {
  try {
    const { image_url, platform, reviewer_name, caption, is_featured, is_featured_on_home } = req.body as {
      image_url?: string;
      platform?: string;
      reviewer_name?: string | null;
      caption?: string | null;
      is_featured?: boolean;
      is_featured_on_home?: boolean;
    };
    if (!image_url) {
      return res.status(400).json({ success: false, error: 'image_url required' });
    }

    const nextOrder = await db.query(
      'SELECT COALESCE(MAX(sort_order), 0) + 1 AS next FROM testimonials',
    );

    const result = await db.query(
      `INSERT INTO testimonials (image_url, platform, reviewer_name, caption, is_featured, is_featured_on_home, sort_order)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       RETURNING *`,
      [
        image_url,
        normalisePlatform(platform),
        reviewer_name ?? null,
        caption ?? null,
        is_featured ?? false,
        is_featured_on_home ?? false,
        nextOrder.rows[0].next,
      ],
    );
    return res.status(201).json({ success: true, testimonial: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// POST /api/admin/testimonials/bulk
// Body (preferred): { items: [{ image_url, platform?, reviewer_name?, caption?, is_featured? }, ...] }
// Body (legacy):    { image_urls: string[] }  — defaults each row to platform='other'
router.post('/bulk', async (req: Request, res: Response) => {
  type BulkItem = {
    image_url?: string;
    platform?: string;
    reviewer_name?: string | null;
    caption?: string | null;
    is_featured?: boolean;
    is_featured_on_home?: boolean;
  };

  const body = req.body as { items?: BulkItem[]; image_urls?: string[] };
  let items: BulkItem[] = [];
  if (Array.isArray(body.items) && body.items.length > 0) {
    items = body.items;
  } else if (Array.isArray(body.image_urls) && body.image_urls.length > 0) {
    items = body.image_urls.map(u => ({ image_url: u }));
  } else {
    return res.status(400).json({
      success: false,
      error: 'items (non-empty array) or image_urls required',
    });
  }

  for (const it of items) {
    if (!it.image_url || typeof it.image_url !== 'string') {
      return res.status(400).json({ success: false, error: 'each item requires image_url' });
    }
  }

  const client = await db.pool.connect();
  try {
    await client.query('BEGIN');
    const baseRow = await client.query(
      'SELECT COALESCE(MAX(sort_order), 0) AS base FROM testimonials',
    );
    let order = parseInt(baseRow.rows[0].base, 10);
    const inserted: unknown[] = [];
    for (const it of items) {
      order += 1;
      const r = await client.query(
        `INSERT INTO testimonials (image_url, platform, reviewer_name, caption, is_featured, is_featured_on_home, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *`,
        [
          it.image_url,
          normalisePlatform(it.platform),
          it.reviewer_name?.trim() ? it.reviewer_name.trim() : null,
          it.caption?.trim()       ? it.caption.trim()       : null,
          Boolean(it.is_featured),
          Boolean(it.is_featured_on_home),
          order,
        ],
      );
      inserted.push(r.rows[0]);
    }
    await client.query('COMMIT');
    return res.status(201).json({ success: true, testimonials: inserted });
  } catch (err) {
    await client.query('ROLLBACK');
    return res.status(500).json({ success: false, error: (err as Error).message });
  } finally {
    client.release();
  }
});

// PUT /api/admin/testimonials/:id
router.put('/:id', async (req: Request, res: Response) => {
  try {
    const { image_url, platform, reviewer_name, caption, is_featured, is_featured_on_home } = req.body as {
      image_url?: string;
      platform?: string;
      reviewer_name?: string | null;
      caption?: string | null;
      is_featured?: boolean;
      is_featured_on_home?: boolean;
    };

    const allowed: Record<string, unknown> = {};
    if (image_url           !== undefined) allowed['image_url']           = image_url;
    if (platform            !== undefined) allowed['platform']            = normalisePlatform(platform);
    if (reviewer_name       !== undefined) allowed['reviewer_name']       = reviewer_name ?? null;
    if (caption             !== undefined) allowed['caption']             = caption       ?? null;
    if (is_featured         !== undefined) allowed['is_featured']         = is_featured;
    if (is_featured_on_home !== undefined) allowed['is_featured_on_home'] = is_featured_on_home;

    if (Object.keys(allowed).length === 0) {
      return res.status(400).json({ success: false, error: 'No fields to update' });
    }

    const setClauses = Object.keys(allowed).map((k, i) => `${k} = $${i + 1}`);
    const values     = [...Object.values(allowed), req.params.id];
    const result = await db.query(
      `UPDATE testimonials SET ${setClauses.join(', ')} WHERE id = $${values.length} RETURNING *`,
      values,
    );
    if (!result.rows.length) return res.status(404).json({ success: false, error: 'Not found' });
    return res.json({ success: true, testimonial: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// PATCH /api/admin/testimonials/reorder
// Body: { items: [{ id, sort_order }, …] }
router.patch('/reorder', async (req: Request, res: Response) => {
  const { items } = req.body as { items?: { id: number; sort_order: number }[] };
  if (!Array.isArray(items)) {
    return res.status(400).json({ success: false, error: 'items array required' });
  }
  const client = await db.pool.connect();
  try {
    await client.query('BEGIN');
    for (const it of items) {
      if (typeof it.id !== 'number' || typeof it.sort_order !== 'number') continue;
      await client.query(
        'UPDATE testimonials SET sort_order = $1 WHERE id = $2',
        [it.sort_order, it.id],
      );
    }
    await client.query('COMMIT');
    return res.json({ success: true });
  } catch (err) {
    await client.query('ROLLBACK');
    return res.status(500).json({ success: false, error: (err as Error).message });
  } finally {
    client.release();
  }
});

// DELETE /api/admin/testimonials/:id — also removes the S3 image (best-effort)
router.delete('/:id', async (req: Request, res: Response) => {
  try {
    const result = await db.query(
      'DELETE FROM testimonials WHERE id = $1 RETURNING id, image_url',
      [req.params.id],
    );
    if (!result.rows.length) return res.status(404).json({ success: false, error: 'Not found' });

    const imageUrl = result.rows[0].image_url as string | null;
    if (imageUrl) {
      try { await deleteFromS3(imageUrl); }
      catch (s3err) { console.warn('[testimonials] S3 cleanup failed:', (s3err as Error).message); }
    }
    return res.json({ success: true, message: 'Testimonial deleted' });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

export default router;

File: backend-golang/internal/handlers/admin/testimonials.go
package admin

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"inanhxink/backend-golang/internal/config"
	"inanhxink/backend-golang/internal/handlers"
)

// GET /api/admin/testimonials
func ListTestimonials(w http.ResponseWriter, r *http.Request) {
	rows, err := config.DB.Query(context.Background(), `
		SELECT * FROM testimonials ORDER BY is_featured DESC, sort_order ASC, created_at DESC`)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	ts, err := handlers.CollectRows(rows)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true, "testimonials": ts})
}

// POST /api/admin/testimonials
func CreateTestimonial(w http.ResponseWriter, r *http.Request) {
	var body struct {
		ImageURL          string  `json:"image_url"`
		ReviewerName      *string `json:"reviewer_name"`
		Caption           *string `json:"caption"`
		IsFeatured        bool    `json:"is_featured"`
		IsFeaturedOnHome  bool    `json:"is_featured_on_home"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.ImageURL == "" {
		handlers.BadRequest(w, "image_url required")
		return
	}

	nextOrder := config.DB.QueryRow(context.Background(),
		"SELECT COALESCE(MAX(sort_order), 0) + 1 AS next FROM testimonials")
	var next int
	nextOrder.Scan(&next) //nolint

	rows, err := config.DB.Query(context.Background(), `
		INSERT INTO testimonials (image_url, reviewer_name, caption, is_featured, is_featured_on_home, sort_order)
		VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
		body.ImageURL, body.ReviewerName, body.Caption, body.IsFeatured, body.IsFeaturedOnHome, next)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, _ := handlers.CollectOne(rows)
	handlers.Created(w, map[string]any{"success": true, "testimonial": row})
}

// POST /api/admin/testimonials/bulk
func BulkCreateTestimonials(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Items     []map[string]any `json:"items"`
		ImageURLs []string         `json:"image_urls"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		handlers.BadRequest(w, "Invalid JSON")
		return
	}

	type bulkItem struct {
		ImageURL         string
		ReviewerName     *string
		Caption          *string
		IsFeatured       bool
		IsFeaturedOnHome bool
	}

	var items []bulkItem
	if len(body.Items) > 0 {
		for _, it := range body.Items {
			url, _ := it["image_url"].(string)
			if url == "" {
				handlers.BadRequest(w, "each item requires image_url")
				return
			}
			bi := bulkItem{ImageURL: url}
			if v, ok := it["reviewer_name"].(string); ok && strings.TrimSpace(v) != "" {
				bi.ReviewerName = &v
			}
			if v, ok := it["caption"].(string); ok && strings.TrimSpace(v) != "" {
				bi.Caption = &v
			}
			bi.IsFeatured, _ = it["is_featured"].(bool)
			bi.IsFeaturedOnHome, _ = it["is_featured_on_home"].(bool)
			items = append(items, bi)
		}
	} else if len(body.ImageURLs) > 0 {
		for _, u := range body.ImageURLs {
			items = append(items, bulkItem{ImageURL: u})
		}
	} else {
		handlers.BadRequest(w, "items (non-empty array) or image_urls required")
		return
	}

	tx, err := config.DB.Begin(context.Background())
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	defer tx.Rollback(context.Background()) //nolint

	var base int
	tx.QueryRow(context.Background(), "SELECT COALESCE(MAX(sort_order), 0) FROM testimonials").Scan(&base) //nolint

	var inserted []map[string]any
	for i, it := range items {
		order := base + i + 1
		rows, err := tx.Query(context.Background(), `
			INSERT INTO testimonials (image_url, reviewer_name, caption, is_featured, is_featured_on_home, sort_order)
			VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
			it.ImageURL, it.ReviewerName, it.Caption, it.IsFeatured, it.IsFeaturedOnHome, order)
		if err != nil {
			handlers.InternalError(w, err)
			return
		}
		row, _ := handlers.CollectOne(rows)
		if row != nil {
			inserted = append(inserted, row)
		}
	}

	if err := tx.Commit(context.Background()); err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.Created(w, map[string]any{"success": true, "testimonials": inserted})
}

// PUT /api/admin/testimonials/:id
func UpdateTestimonial(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var fields map[string]any
	if err := json.NewDecoder(r.Body).Decode(&fields); err != nil {
		handlers.BadRequest(w, "Invalid JSON")
		return
	}

	allowed := []string{"image_url", "reviewer_name", "caption", "is_featured", "is_featured_on_home"}
	setClauses := []string{}
	values := []any{}
	i := 1
	for _, key := range allowed {
		if v, ok := fields[key]; ok {
			setClauses = append(setClauses, fmt.Sprintf("%s = $%d", key, i))
			values = append(values, v)
			i++
		}
	}
	if len(setClauses) == 0 {
		handlers.BadRequest(w, "No fields to update")
		return
	}
	values = append(values, id)

	rows, err := config.DB.Query(context.Background(),
		fmt.Sprintf("UPDATE testimonials SET %s WHERE id = $%d RETURNING *",
			joinClauses(setClauses), len(values)),
		values...)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, err := handlers.CollectOne(rows)
	if err != nil || row == nil {
		handlers.NotFound(w)
		return
	}
	handlers.OK(w, map[string]any{"success": true, "testimonial": row})
}

// PATCH /api/admin/testimonials/reorder
func ReorderTestimonials(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Items []struct {
			ID        int `json:"id"`
			SortOrder int `json:"sort_order"`
		} `json:"items"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || len(body.Items) == 0 {
		handlers.BadRequest(w, "items array required")
		return
	}

	tx, err := config.DB.Begin(context.Background())
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	defer tx.Rollback(context.Background()) //nolint

	for _, it := range body.Items {
		tx.Exec(context.Background(), //nolint
			"UPDATE testimonials SET sort_order = $1 WHERE id = $2", it.SortOrder, it.ID)
	}
	if err := tx.Commit(context.Background()); err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true})
}

// DELETE /api/admin/testimonials/:id — also removes the S3 image (best-effort)
func DeleteTestimonial(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	rows, err := config.DB.Query(context.Background(),
		"DELETE FROM testimonials WHERE id = $1 RETURNING id, image_url", id)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, err := handlers.CollectOne(rows)
	if err != nil || row == nil {
		handlers.NotFound(w)
		return
	}
	if imgURL, ok := row["image_url"].(string); ok && imgURL != "" {
		if _, err := config.DeleteFromS3(imgURL); err != nil {
			// best-effort; log but don't fail
			_ = err
		}
	}
	handlers.OK(w, map[string]any{"success": true, "message": "Testimonial deleted"})
}
